arraycontext/container/traversal.py

```diff
@@ -624,17 +626,40 @@ def _flatten(subary: ArrayOrContainerT) -> None:
                         "This functionality needs to be implemented by the "
                         "array context.") from exc
 
-            result.append(flat_subary)
+            result = [flat_subary]
         else:
+            result = []
             for _, isubary in iterable:
-                _flatten(isubary)
+                result.extend(_flatten(isubary))
 
-    _flatten(ary)
+        return result
 
-    if len(result) == 1:
-        return result[0]
+    def _flatten_without_leaf_class(subary: ArrayOrContainerT) -> Any:
```

Comment:
Not sure this needs to be a separate function. Maybe just bolt this to the end of `flatten`?

Reply:
Made it into a function because it's also called when `type(subary) is leaf_class`, mostly to just not repeat the `if len(result) == 1` part.